### ISSUE TITLE
(assertions): argument to skip bundling of assets

### ISSUE THREAD

**Issue by joel-aws** [aws/aws-cdk#18125]
### Description

When running `assertions.Template.from_stack(stack)`, one cannot skip the bundling of assets during synthesis/generation of the CloudFormation template.

### Use Case

When writing unit-tests that leverage the `assertions` library, assets do not always have to be created to check the synthesized template.  Depending on how many assets are created (and their bundling strategy), the process can be lengthy. This feature-request is for an argument which can disable bundling of assets.

### Proposed Solution

`assertions.Template.from_stack(stack, bundle_assets=False)`

### Other information

I tried doing something like below, to no avail:

```python
@jsii.implements(IAspect)
class MockAssetStaging:
    def __init__(self, fake_stage: AssetStaging):
        self.fake_stage = fake_stage

    def visit(self, node):
        if isinstance(node, AssetStaging):
            node = self.fake_stage

app = App()
core_stack = CoreStack(...)

fake_staged_asset = AssetStaging(core_stack, "fake", source_path=f"./hello_world.py")
Aspects.of(app).add(MockAssetStaging(fake_staged_asset))
```

### Acknowledge

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

**Comment by ChrisSargent** [comment]
there is a bundlingStacks option in the context, in our tests we do something along the lines of:

```ts
import context from './cdk.json'

const context = {
  ...baseContext,
  'aws:cdk:bundling-stacks': [],
};
```

**Comment by joel-aws** [comment]
That's a great idea, @ChrisSargent ! I couldn't get it to work, but am going to keep trying iterations of it.

I searched around a bit and saw https://github.com/aws/aws-cdk/issues/18290#issuecomment-1008956462, which makes me think this wasn't possible?

**Comment by joel-aws** [comment]
@ChrisSargent -- I've hit a wall. Would you mind sharing more of your code (or a link to a repo) so I could see how you set bundlingStacks for your tests?

**Comment by ChrisSargent** [comment]
Hi @joel-aws - sorry, I didn't see this comment until now. Not much more to share but here it is:

```ts
const context = {
  ...baseContext,
  'aws:cdk:bundling-stacks': [],
  stage: 'unitTest',
};
const app = new cdk.App({
  context,
});
```

If definitely works for us.

**Comment by rally25rs** [comment]
the `aws:cdk:bundling-stacks` does not seem to work, nor does it seem to be a documented context option that I could find.
I was poking through the source, and the issue seems to be that the constructor for `new NodejsFunction()` will unconditionally call `Bundling.bundle()`, which ends up always trying to execute some docker commands as a child process.
It seems like the `BundlingOptions` passed to `new NodejsFunction({ bundling: { /* add disable flag here? */ }})` could benefit from some custom bundling command overrides, or a flag to disable bundling.

**Comment by LarsFronius** [comment]
> Hi @joel-aws - sorry, I didn't see this comment until now. Not much more to share but here it is:
> 
> ```ts
> const context = {
>   ...baseContext,
>   'aws:cdk:bundling-stacks': [],
> };
> const app = new cdk.App({
>   context,
> });
> ```
> 
> If definitely works for us.

Thanks, this works well for us!
@rally25rs The feature flag is in https://github.com/aws/aws-cdk/blob/74318c7d22bfc00de9e005f68a0a6aaa58c7db39/packages/%40aws-cdk/cx-api/lib/app.ts#L51 and is then evaluated here https://github.com/aws/aws-cdk/blob/ecedb00ee3a3cfcaa2564a679fa635aff38f32d8/packages/%40aws-cdk/core/lib/stack.ts#L1467 / https://github.com/aws/aws-cdk/blob/813c2f17b6c0f1056ed43a8a93f4cffbe9ae9736/packages/%40aws-cdk/core/lib/asset-staging.ts#L195

**Comment by rally25rs** [comment]
@LarsFronius thanks for the reply. It looks like there is a difference between what I was running into vs the original post.
I think the code you linked and the `aws:cdk:bundling-stacks` config probably works with a `Stack` as used in the original issue;
```
app = App()
core_stack = CoreStack(...)
```
However in my case we were using some `NodejsFunction`s, and constructor for `NodejsFunction` doesn't seem to use that config or asset-staging code.

**Comment by bestickley** [comment]
@rally25rs and @joel-aws , should [NodejsFunction](https://github.com/aws/aws-cdk/blob/v2.90.0/packages/aws-cdk-lib/aws-lambda-nodejs/lib/bundling.ts) use `aws:cdk:bundling-stacks` to not bundle if `cdk deploy ... --exclusively` was called with a stack the current function isn't in? I would think so.

It would be great if there was a property on `Stack` that determined whether the current stack is being bundled or not.

**Comment by onhate** [comment]
```
'aws:cdk:bundling-stacks': [],
```

does not seems to be working anymore on latest version

**Comment by castleadmin** [comment]
I agree, that an asset creation parameter would decrease the test execution time significantly. Furthermore, it would help to create apps where the infrastructure definition lives right beside the application code.

The following dirty piece of code can be used as a workaround:

````
import {CodeConfig} from "aws-cdk-lib/aws-lambda";

jest.mock( "aws-cdk-lib/aws-lambda", () => {
  const module = jest.requireActual<typeof import("aws-cdk-lib/aws-lambda")>( "aws-cdk-lib/aws-lambda");
  module.Code.fromAsset = (path: string): any => ({
    path,
    isInline: false,
    bind: (): CodeConfig => {
      return {
        s3Location: {
          bucketName: 'bucketName',
          objectKey: 'objectKey',
        },
      };
    },
    bindToResource: () => {}
  });

  return module;
});
````

**Comment by CaptainDriftwood** [comment]
Any chance we could see some progress made on this issue? Would be very nice to be able to test constructs without having to bundle them.

**Comment by jusdino** [comment]
> Any chance we could see some progress made on this issue? Would be very nice to be able to test constructs without having to bundle them.

@CaptainDriftwood, for what it's worth, there is a method identified to suppress bundling via context, [as mentioned above](https://github.com/aws/aws-cdk/issues/18125#issuecomment-1099392204)

I've adopted this in my test methods to look something like:
```python
    def test_example(self):
        with open('cdk.json', 'r') as f:
            context = json.load(f)['context']

        # Suppresses lambda bundling for tests
        context['aws:cdk:bundling-stacks'] = []

        app = ExampleApp(context=context)
```

**Comment by CaptainDriftwood** [comment]
> > Any chance we could see some progress made on this issue? Would be very nice to be able to test constructs without having to bundle them.
> 
> @CaptainDriftwood, for what it's worth, there is a method identified to suppress bundling via context, [as mentioned above](https://github.com/aws/aws-cdk/issues/18125#issuecomment-1099392204)
> 
> I've adopted this in my test methods to look something like:
> 
> ```python
>     def test_example(self):
>         with open('cdk.json', 'r') as f:
>             context = json.load(f)['context']
> 
>         # Suppresses lambda bundling for tests
>         context['aws:cdk:bundling-stacks'] = []
> 
>         app = ExampleApp(context=context)
> ```

@jusdino Yes, I saw the workaround as mentioned above (and implemented it in my project with success as a pytest fixture), I was just unsure if this functionality is going to be formally adopted in the `Template` class as mentioned in the Proposed Solution, and if so, when could we expect this issue to be marked in progress. It's been open for a couple years so far.

Nonetheless, shout out to @ChrisSargent for the work around.
